pkg/entities/user.go

```diff
@@ -337,3 +337,85 @@ func (e *Entity) GetRoleByGuildLevelConfig(guildID, userID string) (string, erro
 
 	return "", nil
 }
+
+func (e *Entity) HandleInviteTracker(inviter *discordgo.Member, invitee *discordgo.Member) (*response.HandleInviteHistoryResponse, error) {
+	res := &response.HandleInviteHistoryResponse{}
+
+	if inviter != nil {
+		// create inviter if not exist
+		if err := e.CreateUserIfNotExists(inviter.User.ID, inviter.User.Username); err != nil {
+			e.log.Fields(logger.Fields{"userID": inviter.User.ID, "username": inviter.User.Username}).
+				Error(err, "[Entity][CreateInvite] failed to create user for inviter")
+			return nil, err
+		}
+		if err := e.CreateGuildUserIfNotExists(inviter.GuildID, inviter.User.ID, inviter.Nick); err != nil {
+			e.log.Fields(logger.Fields{"userID": inviter.User.ID, "username": inviter.User.Username}).
+				Error(err, "[Entity][CreateInvite] failed to create guild user for inviter")
+			return nil, err
+		}
+
+		totalInvites, err := e.CountInviteHistoriesByGuildUser(inviter.GuildID, inviter.User.ID)
+		if err != nil {
+			e.log.Fields(logger.Fields{"inviterID": invitee.User.ID, "guildID": inviter.GuildID}).
+				Error(err, "[Entity][CreateInvite] failed to count inviter invites")
+			return nil, err
+		}
+
+		res.InvitesAmount = int(totalInvites)
+		res.InviterID = inviter.User.ID
+		res.IsBot = inviter.User.Bot
+	}
+
+	if invitee != nil {
+		// create invitee if not exist
+		if err := e.CreateUserIfNotExists(invitee.User.ID, invitee.User.Username); err != nil {
+			e.log.Fields(logger.Fields{"userID": invitee.User.ID, "username": invitee.User.Username}).
+				Error(err, "[Entity][CreateInvite] failed to create user for invitee")
+			return nil, err
+		}
+		if err := e.repo.GuildUsers.Create(&model.GuildUser{
```

Comment:
why not use `e.CreateGuildUserIfNotExists()` like above?

Reply:
because invitee may be invited by new user. `e.repo.GuildUsers.Create` will replace field `InvitedBy` of this record (this is caused by `BeforeCreate` of `GuildUser`)